// @0gfoundation/0g-compute-ts-sdk/cli.commonjs/sdk/inference/extractor/speech-to-text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpeechToText = void 0;
const extractor_1 = require("./extractor");
function parseUsage(content) {
    if (!content) {
        return null;
    }
    try {
        const usage = JSON.parse(content);
        return usage && typeof usage === 'object' ? usage : null;
    }
    catch {
        return null;
    }
}
function toCount(value) {
    const num = typeof value === 'string' ? parseInt(value, 10) : value;
    return typeof num === 'number' && !isNaN(num) && num > 0 ? num : 0;
}
function isDurationUsage(usage) {
    const seconds = typeof usage.seconds === 'number' ? usage.seconds : 0;
    return usage.type === 'duration' || (usage.type !== 'tokens' && seconds > 0);
}
/**
 * Billable seconds for duration-mode usage: round to the nearest second with
 * a 1-second floor for any positive duration, matching the broker's
 * billableSeconds() so the cached fee estimate agrees with settlement.
 */
function billableSeconds(usage) {
    const seconds = typeof usage.seconds === 'number' ? usage.seconds : 0;
    if (seconds <= 0) {
        return 0;
    }
    return Math.max(Math.round(seconds), 1);
}
class SpeechToText extends extractor_1.Extractor {
    svcInfo;
    constructor(svcInfo) {
        super();
        this.svcInfo = svcInfo;
    }
    getSvcInfo() {
        return Promise.resolve(this.svcInfo);
    }
    async getInputCount(content) {
        const usage = parseUsage(content);
        if (!usage) {
            return 0;
        }
        if (isDurationUsage(usage)) {
            // Duration billing: inputPrice is price-per-second
            return billableSeconds(usage);
        }
        return toCount(usage.input_tokens);
    }
    async getOutputCount(content) {
        const usage = parseUsage(content);
        if (!usage) {
            return 0;
        }
        if (isDurationUsage(usage)) {
            // Duration billing has no generation-side cost
            return 0;
        }
        return toCount(usage.output_tokens);
    }
}
exports.SpeechToText = SpeechToText;
