// @0gfoundation/0g-compute-ts-sdk/lib.commonjs/fine-tuning/broker/broker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FineTuningBroker = void 0;
exports.createFineTuningBroker = createFineTuningBroker;
const contract_1 = require("../contract");
const model_1 = require("./model");
const dataset_1 = require("./dataset");
const service_1 = require("./service");
const provider_1 = require("../provider/provider");
const utils_1 = require("../../common/utils");
const verifier_1 = require("./verifier");
const read_only_broker_1 = require("./read-only-broker");
class FineTuningBroker extends read_only_broker_1.ReadOnlyFineTuningBroker {
    signer;
    fineTuningCA;
    ledger;
    datasetProcessor;
    verifier;
    serviceProvider;
    _gasPrice;
    _maxGasPrice;
    _step;
    constructor(signer, fineTuningCA, ledger, gasPrice, maxGasPrice, step) {
        super(signer, fineTuningCA);
        this.signer = signer;
        this.fineTuningCA = fineTuningCA;
        this.ledger = ledger;
        this._gasPrice = gasPrice;
        this._maxGasPrice = maxGasPrice;
        this._step = step;
    }
    async initialize() {
        let userAddress;
        try {
            userAddress = await this.signer.getAddress();
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
        const contract = new contract_1.FineTuningServingContract(this.signer, this.fineTuningCA, userAddress, this._gasPrice, this._maxGasPrice, this._step);
        this.serviceProvider = new provider_1.Provider(contract);
        this.modelProcessor = new model_1.ModelProcessor(contract, this.ledger, this.serviceProvider);
        this.datasetProcessor = new dataset_1.DatasetProcessor(contract, this.ledger, this.serviceProvider);
        this.serviceProcessor = new service_1.ServiceProcessor(contract, this.ledger, this.serviceProvider);
        this.verifier = new verifier_1.Verifier(contract, this.ledger, this.serviceProvider);
    }
    getLockedTime = async () => {
        try {
            return await this.serviceProcessor.getLockTime();
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getAccount = async (providerAddress) => {
        try {
            return await this.serviceProcessor.getAccount(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getAccountWithDetail = async (providerAddress) => {
        try {
            return await this.serviceProcessor.getAccountWithDetail(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    acknowledgeProviderSigner = async (providerAddress, gasPrice) => {
        try {
            return await this.serviceProcessor.acknowledgeProviderSigner(providerAddress, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    acknowledgeTEESignerByOwner = async (providerAddress, gasPrice) => {
        try {
            return await this.serviceProcessor.acknowledgeTEESignerByOwner(providerAddress, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    revokeTEESignerAcknowledgement = async (providerAddress, gasPrice) => {
        try {
            return await this.serviceProcessor.revokeTEESignerAcknowledgement(providerAddress, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    removeService = async (gasPrice) => {
        try {
            return await this.serviceProcessor.removeService(gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    modelUsage = (providerAddress, preTrainedModelName, output) => {
        try {
            return this.serviceProcessor.modelUsage(providerAddress, preTrainedModelName, output);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    uploadDataset = async (dataPath, gasPrice, maxGasPrice) => {
        try {
            return await this.datasetProcessor.uploadDataset(this.signer.privateKey, dataPath, gasPrice || this._gasPrice, maxGasPrice || this._maxGasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    downloadDataset = async (dataPath, dataRoot) => {
        try {
            await this.datasetProcessor.downloadDataset(dataPath, dataRoot);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    calculateToken = async (datasetPath, preTrainedModelName, usePython, providerAddress) => {
        try {
            await this.datasetProcessor.calculateToken(datasetPath, usePython, preTrainedModelName, providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    createTask = async (providerAddress, preTrainedModelName, datasetHash, trainingPath, gasPrice) => {
        try {
            return await this.serviceProcessor.createTask(providerAddress, preTrainedModelName, datasetHash, trainingPath, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    cancelTask = async (providerAddress, taskID) => {
        try {
            return await this.serviceProcessor.cancelTask(providerAddress, taskID);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    listTask = async (providerAddress) => {
        try {
            return await this.serviceProcessor.listTask(providerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getTask = async (providerAddress, taskID) => {
        try {
            const task = await this.serviceProcessor.getTask(providerAddress, taskID);
            return task;
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    getLog = async (providerAddress, taskID) => {
        try {
            return await this.serviceProcessor.getLog(providerAddress, taskID);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    acknowledgeModel = async (providerAddress, taskId, dataPath, options) => {
        try {
            return await this.modelProcessor.acknowledgeModel(providerAddress, taskId, dataPath, options);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Acknowledge a delivered task on-chain *without* downloading the model.
     *
     * Escape hatch for the May 2026 hackathon bug report's Bug #4. If a
     * user retrieved the artifact through some out-of-band path (e.g. the
     * legacy `downloadModelFrom0GStorage` + `decryptModel` two-step) and
     * then never called `acknowledgeModel`, every subsequent
     * `addDeliverable` for the same `(user, provider)` pair will revert
     * on-chain with "previous deliverable not acknowledged". After the
     * artifact is garbage-collected from both 0G Storage and TEE buffers,
     * `acknowledgeModel` itself can no longer succeed. Calling
     * `acknowledgeDeliverable(provider, taskId)` directly releases the
     * queue without requiring artifact retrieval.
     *
     * For the normal happy path always prefer {@link FineTuningBroker.acknowledgeModel}
     * — it downloads, verifies the model hash, and acks in one step.
     *
     * @param providerAddress - The on-chain address of the provider that
     *   served the task. Must be the same provider passed to
     *   {@link FineTuningBroker.createTask} for `taskId`.
     * @param taskId - The task identifier returned by
     *   {@link FineTuningBroker.createTask}, in 0x-prefixed hex form.
     * @param gasPrice - Optional gas price (wei) for the
     *   `acknowledgeDeliverable` transaction. When omitted the broker uses
     *   the network-suggested gas price.
     * @returns A promise that resolves once the on-chain transaction has
     *   been mined and the deliverable is acknowledged.
     * @throws An `Error` formatted by `throwFormattedError` if the on-chain
     *   call reverts (e.g. the deliverable is already acknowledged, the
     *   task does not belong to the caller, or the provider has no pending
     *   deliverable for the user).
     */
    acknowledgeDeliverable = async (providerAddress, taskId, gasPrice) => {
        try {
            return await this.modelProcessor.acknowledgeDeliverable(providerAddress, taskId, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Download LoRA model directly from TEE.
     *
     * @param options.idleTimeoutMs - Abort the download if no bytes are
     *   received for this many ms (default 60 000). This is an idle
     *   timeout — slow but live downloads are not killed. Tune up for
     *   networks with long stalls; the broker has no total-request cap
     *   on this path.
     * @param options.maxRetries - Retry attempts on transient stream /
     *   5xx errors (default 2 → 3 attempts total).
     */
    downloadLoRAFromTEE = async (providerAddress, taskId, outputPath, options) => {
        try {
            return await this.modelProcessor.downloadLoRAFromTEE(providerAddress, taskId, outputPath, options);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Upload dataset directly to TEE (broker)
     * Returns the dataset hash for use in task creation
     * This is the preferred method for testing
     */
    uploadDatasetToTEE = async (providerAddress, datasetPath) => {
        try {
            return await this.serviceProvider.uploadDatasetToTEE(providerAddress, datasetPath);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Download encrypted model from 0G Storage (for advanced use cases).
     *
     * @deprecated Prefer {@link FineTuningBroker.acknowledgeModel}, which
     * downloads, verifies the on-chain model hash, and acknowledges the
     * deliverable in a single call. Calling this method on its own leaves
     * the deliverable in an unacknowledged state on-chain — every
     * subsequent `addDeliverable` for the same `(user, provider)` pair will
     * then revert with "previous deliverable not acknowledged",
     * permanently locking the user's task queue. This is the trigger for
     * Bug #4 in the May 2026 hackathon report. If you must use this
     * advanced path, ensure you call
     * {@link FineTuningBroker.acknowledgeDeliverable} (or
     * {@link FineTuningBroker.acknowledgeModel}) afterwards.
     */
    downloadModelFrom0GStorage = async (providerAddress, taskId, dataPath) => {
        try {
            return await this.modelProcessor.downloadModelFrom0GStorage(providerAddress, taskId, dataPath);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Decrypt an already-downloaded encrypted model artifact.
     *
     * @deprecated Prefer {@link FineTuningBroker.acknowledgeModel}; that
     * call downloads, verifies, and acks in one go and never leaves the
     * deliverable in an unacknowledged state. Using `decryptModel` as the
     * second half of the legacy two-step pattern (after
     * `downloadModelFrom0GStorage`) does **not** acknowledge the
     * deliverable on-chain — see {@link FineTuningBroker.downloadModelFrom0GStorage}
     * for the queue-locking failure mode this caused in the May 2026 bug
     * report (Bug #4).
     */
    decryptModel = async (providerAddress, taskId, encryptedModelPath, decryptedModelPath) => {
        try {
            return await this.modelProcessor.decryptModel(providerAddress, taskId, encryptedModelPath, decryptedModelPath);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
    /**
     * Verify fine-tuning service TEE attestation (DStack only)
     *
     * Downloads and verifies the TEE attestation report to ensure the provider
     * is running in a trusted execution environment. This simplified version
     * only supports DStack (Intel TDX) verification.
     *
     * @param providerAddress - The provider address to verify
     * @param outputDir - Directory to save attestation reports (default: current directory)
     * @returns Promise resolving to verification results with report paths and status
     *
     * @example
     * ```typescript
     * const broker = await createFineTuningBroker(signer, contractAddress, ledger);
     *
     * // Verify a provider's TEE attestation
     * const result = await broker.verifyService(
     *   '0x1234...',
     *   './attestation-reports'
     * );
     *
     * if (result.success && result.reportsData) {
     *   console.log('Verification successful');
     *   console.log('Reports saved to:', result.outputDirectory);
     * }
     * ```
     *
     * @remarks
     * This method downloads the attestation report and performs automated checks:
     * 1. TEE Signer Address Verification - matches contract signer with report signer
     * 2. Docker Compose Verification - validates compose hash against event log
     *
     * After automated checks, users must manually verify:
     * 1. Docker images using sigstore (https://search.sigstore.dev/)
     * 2. Run dstack-verifier for complete quote verification
     *
     * @throws {Error} If provider doesn't exist or attestation report cannot be retrieved
     */
    verifyService = async (providerAddress, outputDir = '.', onLog) => {
        try {
            return await this.verifier.verifyService(providerAddress, outputDir, onLog);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    };
}
exports.FineTuningBroker = FineTuningBroker;
/**
 * createFineTuningBroker is used to initialize ZGServingUserBroker
 *
 * @param signer - Signer from ethers.js.
 * @param contractAddress - 0G Serving contract address, use default address if not provided.
 * @param ledger - Ledger broker instance.
 * @param gasPrice - Gas price for transactions. If not provided, the gas price will be calculated automatically.
 *
 * @returns broker instance.
 *
 * @throws An error if the broker cannot be initialized.
 */
async function createFineTuningBroker(signer, contractAddress, ledger, gasPrice, maxGasPrice, step) {
    const broker = new FineTuningBroker(signer, contractAddress, ledger, gasPrice, maxGasPrice, step);
    try {
        await broker.initialize();
        return broker;
    }
    catch (error) {
        throw error;
    }
}
