// @0gfoundation/0g-compute-ts-sdk/lib.commonjs/fine-tuning/broker/model.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelProcessor = void 0;
const utils_1 = require("../../common/utils");
const const_1 = require("../const");
const zg_storage_1 = require("../zg-storage");
const read_only_model_1 = require("./read-only-model");
const logger_1 = require("../../common/logger");
const ethers_1 = require("ethers");
const constants_1 = require("../../constants");
// `fs/promises` and `path` are loaded lazily inside the async helpers below;
// see the comment in src.ts/sdk/fine-tuning/provider/provider.ts for the
// rationale (browser bundlers statically validate top-level Node-builtin
// imports against the package.json `"browser"` empty stub).
/**
 * ModelProcessor handles model-related operations including listing available models,
 * acknowledging model delivery, and decrypting fine-tuned models.
 *
 * Extends ReadOnlyModelProcessor to inherit listModel() for read-only operations.
 */
class ModelProcessor extends read_only_model_1.ReadOnlyModelProcessor {
    ledger;
    servingProvider;
    constructor(contract, ledger, servingProvider) {
        super(contract);
        this.contract = contract;
        this.ledger = ledger;
        this.servingProvider = servingProvider;
    }
    /**
     * Get storage configuration based on current network
     */
    async getStorageConfig() {
        try {
            const chainId = await this.contract.signer.provider
                ?.getNetwork()
                .then((n) => n.chainId);
            const networkType = chainId ? (0, constants_1.getNetworkType)(chainId) : 'unknown';
            if (networkType === 'mainnet') {
                return {
                    rpcUrl: const_1.ZG_RPC_ENDPOINT_MAINNET,
                    indexerUrl: const_1.INDEXER_URL_MAINNET_TURBO,
                };
            }
            else {
                return {
                    rpcUrl: const_1.ZG_RPC_ENDPOINT_TESTNET,
                    indexerUrl: const_1.INDEXER_URL_TESTNET_TURBO,
                };
            }
        }
        catch {
            return {
                rpcUrl: const_1.ZG_RPC_ENDPOINT_TESTNET,
                indexerUrl: const_1.INDEXER_URL_TESTNET_TURBO,
            };
        }
    }
    /**
     * Acknowledge model delivery and download the trained model
     * @param providerAddress - The provider's address
     * @param taskId - The task ID
     * @param dataPath - Path to save the downloaded model
     * @param options - Optional configuration
     * @param options.gasPrice - Gas price for the transaction
     * @param options.downloadMethod - Download method: 'auto' (default, try 0G Storage first then TEE fallback), 'tee', or '0g-storage'
     * @param options.teeIdleTimeoutMs - For the TEE path: abort if no bytes
     *   are received for this many ms (default 60 000). This is an *idle*
     *   timeout, not a total-request timeout — large models stream fine even
     *   when they take well over 5 minutes.
     * @param options.teeMaxRetries - For the TEE path: number of retry
     *   attempts on transient stream / 5xx errors (default 2 → 3 attempts
     *   total).
     */
    async acknowledgeModel(providerAddress, taskId, dataPath, options) {
        try {
            const fs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            const path = await Promise.resolve().then(() => __importStar(require('path')));
            const gasPrice = options?.gasPrice;
            const downloadMethod = options?.downloadMethod ?? 'auto';
            const teeDownloadOptions = {
                idleTimeoutMs: options?.teeIdleTimeoutMs,
                maxRetries: options?.teeMaxRetries,
            };
            const deliverable = await this.contract.getDeliverable(providerAddress, taskId);
            logger_1.logger.debug(`deliverable: ${deliverable.modelRootHash}`);
            if (!deliverable) {
                throw new Error('No deliverable found');
            }
            // Resolve storage download path: 0G Storage client needs a file path, not a directory
            let storageDownloadPath = dataPath;
            try {
                const stats = await fs.stat(dataPath);
                if (stats.isDirectory()) {
                    storageDownloadPath = path.join(dataPath, `model_${taskId}.bin`);
                }
            }
            catch {
                // Path doesn't exist yet, use as-is (will be created as a file)
            }
            if (downloadMethod === 'tee') {
                // Download LoRA directly from TEE
                await this.servingProvider.downloadLoRAFromTEE(providerAddress, taskId, dataPath, teeDownloadOptions);
                logger_1.logger.info('Successfully downloaded LoRA model from TEE');
                // Verify hash of downloaded file against on-chain modelRootHash
                await this.verifyDownloadedModelHash(dataPath, taskId, deliverable.modelRootHash);
            }
            else if (downloadMethod === '0g-storage') {
                // Download from 0G Storage with built-in hash verification
                const storageConfig = await this.getStorageConfig();
                await (0, zg_storage_1.download)(storageDownloadPath, deliverable.modelRootHash, storageConfig);
                logger_1.logger.info(`Successfully downloaded model from 0G Storage to ${storageDownloadPath}`);
            }
            else {
                // Auto mode: try 0G Storage first, fallback to TEE
                try {
                    logger_1.logger.info('Downloading model from 0G Storage...');
                    const storageConfig = await this.getStorageConfig();
                    await (0, zg_storage_1.download)(storageDownloadPath, deliverable.modelRootHash, storageConfig);
                    logger_1.logger.info(`Successfully downloaded model from 0G Storage to ${storageDownloadPath}`);
                }
                catch (storageErr) {
                    logger_1.logger.warn(`0G Storage download failed: ${storageErr}. Falling back to TEE download...`);
                    await this.servingProvider.downloadLoRAFromTEE(providerAddress, taskId, dataPath, teeDownloadOptions);
                    logger_1.logger.info('Successfully downloaded LoRA model from TEE (fallback)');
                    // Verify hash of downloaded file against on-chain modelRootHash
                    await this.verifyDownloadedModelHash(dataPath, taskId, deliverable.modelRootHash);
                }
            }
            await this.contract.acknowledgeDeliverable(providerAddress, taskId, gasPrice);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    /**
     * Acknowledge a delivered task on-chain *without* downloading or
     * verifying the model artifact.
     *
     * This is the escape hatch for the failure mode reported in the May
     * 2026 hackathon bug report (Bug #4): a user retrieved a model via
     * the legacy two-step `downloadModelFrom0GStorage` + `decryptModel`
     * flow and forgot to call `acknowledgeModel`. Days later the artifact
     * was garbage-collected from both 0G Storage and the TEE buffer, at
     * which point `acknowledgeModel` could no longer succeed (it requires
     * a successful download), and the user's deliverable queue was
     * permanently locked — every subsequent `addDeliverable` reverted with
     * "previous deliverable not acknowledged".
     *
     * Calling this method directly with the stuck task id releases the
     * queue without requiring any artifact retrieval.
     *
     * Prefer `acknowledgeModel(...)` for the normal happy path; that
     * function downloads, verifies the artifact hash, and then acks. Use
     * `acknowledgeDeliverable` only when:
     *   - the artifact is gone or you have already retrieved it offline, AND
     *   - you accept that no on-chain hash verification is performed.
     *
     * @param providerAddress - Address of the provider who delivered the task
     * @param taskId - The task id whose deliverable is to be acknowledged
     * @param gasPrice - Optional gas price override for the on-chain transaction
     */
    async acknowledgeDeliverable(providerAddress, taskId, gasPrice) {
        try {
            const deliverable = await this.contract.getDeliverable(providerAddress, taskId);
            if (!deliverable) {
                throw new Error(`No deliverable found for task ${taskId}`);
            }
            if (deliverable.acknowledged) {
                logger_1.logger.info(`Deliverable for task ${taskId} is already acknowledged on-chain — nothing to do.`);
                return;
            }
            await this.contract.acknowledgeDeliverable(providerAddress, taskId, gasPrice);
            logger_1.logger.info(`Acknowledged deliverable for task ${taskId} (provider ${providerAddress}). ` +
                'Note: no model hash was verified by this call. ' +
                'If you have not yet retrieved the artifact, call acknowledgeModel(...) instead.');
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    /**
     * Download the encrypted model file from 0G Storage.
     *
     * @deprecated For the normal retrieval flow prefer
     * {@link ModelProcessor.acknowledgeModel} which downloads, verifies the
     * on-chain model hash, and acknowledges the deliverable in one call.
     *
     * Calling this method on its own (without a subsequent
     * {@link ModelProcessor.acknowledgeDeliverable} or
     * {@link ModelProcessor.acknowledgeModel}) leaves the deliverable in an
     * "unacknowledged" state on-chain. The provider contract then rejects
     * any future `addDeliverable` for the same `(user, provider)` pair
     * with "previous deliverable not acknowledged", permanently locking
     * the user's queue. This is the May 2026 bug report's Bug #4 trigger.
     *
     * Use this method only as part of an advanced retrieval pipeline where
     * you control acknowledgement separately, and remember to call
     * `acknowledgeDeliverable(provider, taskId)` afterwards.
     */
    async downloadModelFrom0GStorage(providerAddress, taskId, dataPath) {
        try {
            const fs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            const path = await Promise.resolve().then(() => __importStar(require('path')));
            const deliverable = await this.contract.getDeliverable(providerAddress, taskId);
            if (!deliverable) {
                throw new Error('No deliverable found');
            }
            // Resolve path: 0G Storage client needs a file path, not a directory
            let downloadPath = dataPath;
            try {
                const stats = await fs.stat(dataPath);
                if (stats.isDirectory()) {
                    downloadPath = path.join(dataPath, `model_${taskId}.bin`);
                }
            }
            catch {
                // Path doesn't exist yet, use as-is
            }
            const storageConfig = await this.getStorageConfig();
            await (0, zg_storage_1.download)(downloadPath, deliverable.modelRootHash, storageConfig);
            logger_1.logger.info(`Successfully downloaded model from 0G Storage to ${downloadPath}`);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    /**
     * Download LoRA model directly from TEE (without acknowledge).
     * Use this when you only want to download the trained LoRA adapter.
     *
     * @param options.idleTimeoutMs - Abort the download if no bytes are
     *   received for this many ms (default 60 000). This is an idle
     *   timeout — slow but live downloads are not killed.
     * @param options.maxRetries - Retry attempts on transient stream /
     *   5xx errors (default 2 → 3 attempts total).
     */
    async downloadLoRAFromTEE(providerAddress, taskId, outputPath, options) {
        try {
            await this.servingProvider.downloadLoRAFromTEE(providerAddress, taskId, outputPath, options);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
    }
    /**
     * Verify the hash of a downloaded model file against the expected on-chain hash.
     */
    async verifyDownloadedModelHash(filePath, taskId, expectedHash) {
        try {
            const fs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            let actualFile = filePath;
            try {
                const stats = await fs.stat(filePath);
                if (stats.isDirectory()) {
                    const files = await fs.readdir(filePath);
                    const loraFile = files.find((f) => f.startsWith('lora_model_') ||
                        f.endsWith('.data') ||
                        f.endsWith('.zip'));
                    if (loraFile) {
                        actualFile = `${filePath}/${loraFile}`;
                    }
                    else if (files.length === 1) {
                        actualFile = `${filePath}/${files[0]}`;
                    }
                    else {
                        logger_1.logger.warn(`Cannot determine downloaded file in directory ${filePath}, skipping hash verification`);
                        return;
                    }
                }
            }
            catch (err) {
                logger_1.logger.warn(`Downloaded file not found at ${filePath}, skipping hash verification`);
                return;
            }
            const fileData = await fs.readFile(actualFile);
            const computedHash = ethers_1.ethers.keccak256(fileData);
            if (expectedHash &&
                expectedHash !==
                    '0x0000000000000000000000000000000000000000000000000000000000000000') {
                if (computedHash !== expectedHash) {
                    logger_1.logger.warn(`Hash mismatch for task ${taskId}: expected ${expectedHash}, got ${computedHash}`);
                }
                else {
                    logger_1.logger.info(`Hash verification passed for task ${taskId}`);
                }
            }
            else {
                logger_1.logger.info(`No on-chain hash to verify against for task ${taskId}, computed hash: ${computedHash}`);
            }
        }
        catch (err) {
            logger_1.logger.warn(`Hash verification failed: ${err}`);
        }
    }
    /**
     * Decrypt a fine-tuned model after acknowledgement.
     * Uses the user's private key to decrypt the model encryption key,
     * then decrypts the model file.
     *
     * @param providerAddress - Address of the provider who trained the model
     * @param taskId - ID of the fine-tuning task
     * @param encryptedModelPath - Local path to the encrypted model file
     * @param decryptedModelPath - Local path where the decrypted model will be saved
     * @throws Error if deliverable not found, not acknowledged, or decryption fails
     *
     * @example
     * ```typescript
     * await broker.fineTuning.decryptModel(
     *   '0x1234...',
     *   'task-123',
     *   './encrypted-model.bin',
     *   './my-model'
     * );
     * ```
     *
     * @deprecated For the normal retrieval flow prefer
     * {@link ModelProcessor.acknowledgeModel} which downloads, verifies, and
     * acknowledges in one call. `decryptModel` is the second half of the
     * legacy two-step pattern (`downloadModelFrom0GStorage` + `decryptModel`)
     * — neither half acknowledges the deliverable on-chain. Forgetting the
     * acknowledgement is what triggered Bug #4 in the May 2026 hackathon
     * bug report, where a user's task queue became permanently locked.
     *
     * @remarks
     * The model can only be decrypted after:
     * 1. The provider has delivered the encrypted model
     * 2. The user has acknowledged the model (called acknowledgeModel
     *    OR acknowledgeDeliverable)
     * 3. The provider has shared the encrypted decryption key
     */
    async decryptModel(providerAddress, taskId, encryptedModelPath, decryptedModelPath) {
        try {
            const [service, deliverable] = await Promise.all([
                this.contract.getService(providerAddress),
                this.contract.getDeliverable(providerAddress, taskId),
            ]);
            logger_1.logger.debug(`service, ${service}`);
            if (!deliverable) {
                throw new Error('No deliverable found');
            }
            if (!deliverable.acknowledged) {
                throw new Error(`Deliverable for task ${taskId} is not acknowledged yet. ` +
                    'Call broker.fineTuning.acknowledgeModel(provider, taskId, dataPath) ' +
                    '(preferred — downloads, verifies, and acks) ' +
                    'or broker.fineTuning.acknowledgeDeliverable(provider, taskId) ' +
                    '(if the artifact is no longer retrievable). ' +
                    'Without acknowledgement the user queue stays locked and any future ' +
                    'fine-tune task will fail with "previous deliverable not acknowledged".');
            }
            if (!deliverable.encryptedSecret) {
                throw new Error('EncryptedSecret not found');
            }
            const secret = await (0, utils_1.eciesDecrypt)(this.contract.signer, deliverable.encryptedSecret);
            await (0, utils_1.aesGCMDecryptToFile)(secret, encryptedModelPath, decryptedModelPath, service.teeSignerAddress);
        }
        catch (error) {
            (0, utils_1.throwFormattedError)(error);
        }
        return;
    }
}
exports.ModelProcessor = ModelProcessor;
